Gems/Compression/Code/Source/Tools/CompressionFactoryImpl.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/std/ranges/ranges_algorithm.h>
+#include "CompressionFactoryImpl.h"
+
+namespace Compression
+{
+    CompressionFactoryImpl::~CompressionFactoryImpl()
+    {
+        for (ICompressionInterface* compressionInterface : m_compressionInterfaces)
+        {
+            delete compressionInterface;
+        }
+    }
+    AZStd::span<ICompressionInterface* const> CompressionFactoryImpl::GetCompressionInterfaces()
+    {
+        return m_compressionInterfaces;
+    }
+
+    bool CompressionFactoryImpl::RegisterCompressionInterface(AZStd::unique_ptr<ICompressionInterface>&& compressionInterface)
+    {
+        if (compressionInterface == nullptr)
+        {
+            return false;
+        }
+        auto FindCompression = [compressionAlgorithmId = compressionInterface->GetCompressionAlgorithmId()]
+        (const ICompressionInterface* registeredInterface)
+        {
+            return registeredInterface->GetCompressionAlgorithmId() == compressionAlgorithmId;
+        };
+        if (auto compressionFoundIt = AZStd::ranges::find_if(m_compressionInterfaces, FindCompression);
+            compressionFoundIt != m_compressionInterfaces.end())
+        {
+            return false;
+        }
+
+        // The Compression Interface isn't registered, so move it into the container
+        m_compressionInterfaces.emplace_back(compressionInterface.release());
+        return true;
+
+    }
+
+    bool CompressionFactoryImpl::UnregisterCompressionInterface(CompressionAlgorithmId compressionAlgorithmId)
+    {
+        auto FindCompression = [compressionAlgorithmId]
+        (ICompressionInterface* registeredInterface)
+        {
+            if (registeredInterface->GetCompressionAlgorithmId() == compressionAlgorithmId)
+            {
+                delete registeredInterface;
+                return true;
+            }
+
+            return false;
+        };
+
+        return AZStd::erase_if(m_compressionInterfaces, FindCompression) != 0;
+    }
+
+    ICompressionInterface* CompressionFactoryImpl::FindCompressionInterface(CompressionAlgorithmId compressionAlgorithmId) const
+    {
+        auto FindCompression = [compressionAlgorithmId]
+        (const ICompressionInterface* registeredInterface)
+        {
+            return registeredInterface->GetCompressionAlgorithmId() == compressionAlgorithmId;
+        };
+
+        auto compressionFoundIt = AZStd::ranges::find_if(m_compressionInterfaces, FindCompression);
+        return compressionFoundIt != m_compressionInterfaces.end() ? *compressionFoundIt : nullptr;
```

Comment:
Since adding/removing compression interfaces is going to a very rare compared to retrieving an interface, consider keeping the list of compression interfaces sorted and do a binary search. For further improvements consider storing the interface id and the actual data separate as that would be more cache friendly for searches.

Reply:
We could just have a parallel index vector with the compression interface id that is sorted. 
It would just contain an index into the actual vector.

Reply:
Circularing back to this.
Keeping the data separate for cache friendliness was a bad idea.

It turns removal of a compression interface into a `O(n log n)` algorithm as the remove algorithm not only needs to use binary search to locate the element to remove in the compression interface array which is `O(log n).
It also has to iterate over the compression index array linearly to update any indices.

Furthermore the elements being stored in the array have a size 16 bytes. So for a 64 byte cache line 4 elements can fit within the cache, which is more than enough considering that only 7 total compression algorithms are supported

This was a premature optimization gone wrong.
Sometimes we should just start off with the simplest algorithm(an unordered_map) and then attempt to improve performance if there are any issues.

This death by a thousand cuts arguments is a fallacy that developers need to grow out of.